src/chainstate/coordinator/comm.rs

```diff
@@ -0,0 +1,243 @@
+use std::{
+    thread, process
+};
+use std::time::{
+    Duration, Instant
+};
+use std::sync::{
+    Arc, RwLock,
+    atomic::{Ordering, AtomicU64}
+};
+
+use crossbeam_channel::{bounded, Sender, Receiver, TrySendError};
+
+
+/// Trait for use by the ChainsCoordinator
+/// 
+pub trait CoordinatorNotices {
+    fn notify_stacks_block_processed(&mut self);
+    fn notify_sortition_processed(&mut self);
+}
+
+pub struct ArcCounterCoordinatorNotices {
+    pub stacks_blocks_processed: Arc<AtomicU64>,
+    pub sortitions_processed: Arc<AtomicU64>
+}
+
+impl CoordinatorNotices for () {
+    fn notify_stacks_block_processed(&mut self) {}
+    fn notify_sortition_processed(&mut self) {}
+}
+
+impl CoordinatorNotices for ArcCounterCoordinatorNotices {
+    fn notify_stacks_block_processed(&mut self) {
+        self.stacks_blocks_processed.fetch_add(1, Ordering::SeqCst);
+    }
+    fn notify_sortition_processed(&mut self) {
+        self.sortitions_processed.fetch_add(1, Ordering::SeqCst);
+    }
+}
+
+/// Structure used for communication _with_ a running
+///   ChainsCoordinator
+#[derive(Clone)]
+pub struct CoordinatorChannels {
+    // ChainsCoordinator takes two kinds of signals:
+    //    new stacks block & new burn block
+    // These signals can be coalesced -- the coordinator doesn't need
+    //    handles _all_ new blocks whenever it processes an event
+    //    because of this, we can avoid trying to set large bounds on these
+    //    event channels by using a coalescing thread.
+    new_stacks_block_channel: Sender<()>,
+    new_burn_block_channel: Sender<()>,
+    /// how many stacks blocks have been processed by this Coordinator thread since startup?
+    stacks_blocks_processed: Arc<AtomicU64>,
+    /// how many sortitions have been processed by this Coordinator thread since startup?
+    sortitions_processed: Arc<AtomicU64>,
+    stop: Sender<()>
+}
+
+/// Structure used by the Coordinator's run-loop
+///   to receive signals
+pub struct CoordinatorReceivers {
+    pub event_stacks_block: Receiver<()>,
+    pub event_burn_block: Receiver<()>,
+    pub stop: Receiver<()>,
+    pub stacks_blocks_processed: Arc<AtomicU64>,
+    pub sortitions_processed: Arc<AtomicU64>,
+}
+
+// Singletons for ChainsCoordinator communication
+//
+//  these channels allow any thread to notify the ChainsCoordinator
+//   instance that a new staging block is ready or a new bitcoin
+//   block has arrived
+//
+//  using a singleton for this pretty dramatically simplifies state
+//   management in the stacks-node, bitcoin indexer, and relayer, because they
+//   don't need to pass around instances of the channels. however,
+//   this _does_ step on the cargo test framework in silly ways, so any
+//   tests which instantiate a coordinator need to call
+//   CoordinatorCommunication::stop_chains_coordinator()
+//   when they are done.
```

Comment:
This sounds like a pretty bad trade-off.  What state simplification does having a global like this save us?  If all it is doing is saving us from having to fill in a struct field, or have an extra argument, then I don't think having a global is worth it.  Making sure resources get freed when they're supposed to (e.g. managed through object lifetimes) is more important than cutting down on the number of struct fields or function arguments.

Reply:
I think this is _especially_ true if the `blockstack_lib` module is going to be its own library at some point (as it seems to be right now)..  Global mutable state is a show-stopper for that use-case.

Reply:
When would a node not have a coordinator running? The singleton implementation would have fewer resources used, rather than more, since it wouldn't need to ever clone the sender channels.

Reply:
We can certainly move away from using a singleton for the sender channels, though, it would just mean that `Burnchain::sync_with_indexer` and `Relayer::process_network_result` would need to be supplied with references to a `CoordinatorCommunication` object.

Reply:
> When would a node not have a coordinator running?

Many tools -- including miner binaries, chainstate inspectors, MARF inspectors, etc. -- that link against the `blockstack_lib` crate would never instantiate a node.  I don't think we want to force tool implementors to go through an unrelated (and potentially expensive) song-and-dance routine of setting up and tearing down global state that they'll never use.

Also, what happens if the binary needs to instantiate multiple chain coordinators?  The test framework would need this, for example, especially when doing unit tests.

> We can certainly move away from using a singleton for the sender channels, though, it would just mean that Burnchain::sync_with_indexer and Relayer::process_network_result would need to be supplied with references to a CoordinatorCommunication object.

I think this is far more preferable, and a very small cost to pay for avoiding all the future pain of having to explicitly initialize and tear down global state.

Reply:
Singleton no more: d1a5c03

Reply:
:tada: